### ISSUE TITLE
Qt HTMLEditor does not open link when HTML/user gesture opening links externally

### ISSUE THREAD

**Issue by kitchoi** [enthought/traitsui#1464]
Qt WebKit/WebEngine supports opening a link in a new window when the HTML author instructs it to do so with `target='_blank'`:
```
<a href='https://example.com' target='_blank'>Target to _blank Link</a>
```
The user can also elect to open a link in a new window by selecting "Open Link in New Window" from the context menu. 

Unexpected behaviour: HTMLEditor does not open the link at all for the above gestures.
Expected behaviour: HTMLEditor will honour the above gestures and allow the link to be open in a new window. What is a "new window" depends on the context. But given the existing behaviour of the `open_externally` flag on the HTMLEditor is to open a link in the user's machine default web browser, I'd expect that new window to mean the same.

To demonstrate with this example code:

<details>
    <summary> Code snippet </summary>

```

from traits.api import HasTraits, HTML
from traitsui.api import UItem, View, HTMLEditor


html = """
    <h1><a href='https://example.com'>Target to default Link</a></h1>
    <h1><a href='https://example.com' target='_blank'>Target to _blank Link</a></h1>

"""

class HTMLEditorDemo(HasTraits):

    my_html_trait = HTML(html)

    traits_view = View(
        UItem(
            'my_html_trait',
            editor=HTMLEditor(format_text=True)
        ),
        title='HTMLEditor',
        buttons=['OK'],
        width=800,
        height=600,
        resizable=True
    )


# Create the demo:
demo = HTMLEditorDemo()

# Run the demo (if invoked from the command line):
if __name__ == '__main__':
    demo.configure_traits()

```

</details>

This is the current behaviour (video is created using PySide2 + WebEngine, but PyQt5 + WebKit behaves in pretty much the same way except the "Back" option is not available from the context menu):

https://user-images.githubusercontent.com/3673984/103654983-89105080-4f5e-11eb-9182-1dac2d38a9e6.mp4

If I add `open_externally=True` to `HTMLEditor`, like this:
```
        UItem(
            'my_html_trait',
            editor=HTMLEditor(format_text=True, open_externally=True)
        ),
```

Then all that changes is that the link is open in the default web browser when I click the link without `target='_blank'`, gestures to explicitly open the link in a new window still results in the link not being opened at all:

https://user-images.githubusercontent.com/3673984/103655265-e5737000-4f5e-11eb-98c0-bf48e5559c31.mp4

**Comment by kitchoi** [comment]
I do have a draft implementation to achieve the expected behaviour:

<details>
    <summary> Draft implementation of Qt HTMLEditor </summary>


```

def _is_web_engine_page(page):
    return not hasattr(page, "setLinkDelegationPolicy")


class _OpenLinkExternallyPage(QtWebKit.QWebPage):
    """ This page is for supporting the open_externally trait on the
    HTMLEditor. It makes all links to be opened in the user's web browser
    regardless of user gestures or HTML specification.
    """

    def createWindow(self, type):
        return _PageToWebBrowser(self.parent())

    def acceptNavigationRequest(
            self, frame_or_url, request_or_type, type_or_is_main_frame):

        if _is_web_engine_page(self):
            url = frame_or_url
            type = request_or_type
        else:
            url = request_or_type.url()
            type = type_or_is_main_frame

        if type == QtWebKit.QWebPage.NavigationTypeLinkClicked:
            webbrowser.open_new(url.toString())
            return False

        return super().acceptNavigationRequest(
            frame_or_url, request_or_type, type_or_is_main_frame
        )


class _PageToWebBrowser(QtWebKit.QWebPage):
    """ A page that redirects all navigation to the user's web browser.
    """

    def acceptNavigationRequest(self, frame_or_url, request_or_type, _):
        if not _is_web_engine_page(self):
            url = request_or_type.url()
        else:
            url = frame_or_url
        webbrowser.open_new(url.toString())
        return False


class _CustomWebPage(QtWebKit.QWebPage):
    """ This page redirects all requests to creating new windows to the page
    that redirects all URL to the web browser.
    """

    def createWindow(self, type):
        return _PageToWebBrowser(self.parent())


class SimpleEditor(Editor):
    """ Simple style editor for HTML.
    """

    # -------------------------------------------------------------------------
    #  Trait definitions:
    # -------------------------------------------------------------------------

    #: Is the HTML editor scrollable? This values override the default.
    scrollable = True

    #: External objects referenced in the HTML are relative to this URL
    base_url = Str()

    def init(self, parent):
        """ Finishes initializing the editor by creating the underlying toolkit
            widget.
        """
        self.control = QtWebKit.QWebView()
        self.control.setSizePolicy(
            QtGui.QSizePolicy.Expanding, QtGui.QSizePolicy.Expanding
        )
        if self.factory.open_externally:
            self.control.setPage(_OpenLinkExternallyPage(self.control))
        else:
            self.control.setPage(_CustomWebPage(self.control))

        self.base_url = self.factory.base_url
        self.sync_value(self.factory.base_url_name, "base_url", "from")

    def dispose(self):
        """ Disposes of the contents of an editor.
        """
        super().dispose()

    def update_editor(self):
        """ Updates the editor when the object trait changes external to the
            editor.
        """
        text = self.str_value
        if self.factory.format_text:
            text = self.factory.parse_text(text)
        if self.base_url:
            url = self.base_url
            if not url.endswith("/"):
                url += "/"
            self.control.setHtml(text, QtCore.QUrl.fromLocalFile(url))
        else:
            self.control.setHtml(text)

    # -- Event Handlers -------------------------------------------------------

    def _base_url_changed(self):
        self.update_editor()

```

</details>


And this is the resulting behaviour (with `open_externally` set to False):

https://user-images.githubusercontent.com/3673984/103656460-86165f80-4f60-11eb-9944-03428b76fd26.mp4

**Comment by kitchoi** [comment]
My observation is that the `open_externally` flag is motivated by the limitation in wx.  The Wx implementation of HTMLEditor uses `wx.html.HtmlWindow` which has limited HTML rendering support: Clicking a link with `open_externally=False` simply does nothing.

However by changing just the following line,
https://github.com/enthought/traitsui/blob/5a8f5d4f80ccadb0cdd94ccdb28183ca35dbb432/traitsui/wx/html_editor.py#L52
to
```
if self.open_externally or link_info.GetTarget() == "_blank":
```
The wx HTMLEditor can honour the `target` setting as well even with `open_externally` set to False.

Therefore the question related to this issue is this: Does `open_externally=False` mean "forbidding attempts to open a link externally"?  I doubt that was the intention based on this commit: https://github.com/enthought/envisage/commit/ac7e5f3af617a19d2b87659851ee5348c5295ae4

**Comment by kitchoi** [comment]
> If I add open_externally=True to HTMLEditor, like this:
> ...
> Then all that changes is that the link is open in the default web browser when I click the link without target='_blank', gestures to explicitly open the link in a new window still results in the link not being opened at all

This description is not quite accurate.
On master, if `open_externally is True`, when one has QtWebEngine (e.g. PySide2 from PyPI), then clicking the link with `target='_blank'` does nothing.  When one has QtWebKit (e.g. PyQt5 from EDM), then clicking the link with `target='_blank'` will open the link in a default web browser.

The draft implementation still does the expected thing.

**Comment by corranwebster** [comment]
> Therefore the question related to this issue is this: Does `open_externally=False` mean "forbidding attempts to open a link externally"? I doubt that was the intention based on this commit: [enthought/envisage@ac7e5f3](https://github.com/enthought/envisage/commit/ac7e5f3af617a19d2b87659851ee5348c5295ae4)

No, False means "do the default thing" whereas True means always open a browser window.  

If it's not working as expected on Wx it's likely a bug, probably from changes in WxPython 4.0, or from the fact that `_blank` was deprecated in HTML4/XHTML which would have been the time when that code was written.

Even now, the use of `_blank` seems controversial amongst web developers, so I would be fine with not supporting it.

From the perspective of scientific applications, there are two use-cases that we really want to support:

- using HTML to display a collection of related documents (eg. help pages or a report) with relative links, but no external links, limited or no javascript, etc. so we want to support navigation capabilities at least in a limited way; and
- using HTML as a convenient "static" rich text format etc. where we don't want to navigate away from the displayed page, so links should go to a full browser.

This is really what the flag is toggling between.

**Comment by kitchoi** [comment]
In addition to `_blank` which is a specification on the part of the HTML author, there is also the user gesture to opening a link in a new window: The user explicitly requests it.  I think that part ought to be honoured right?

In my opinion, I think allowing the user to choose is more friendly than dictating that links must be opened in a default browser (TBH, it is a bit rude :) ). Likewise, HTMLEditor can leave that option open for developers to choose that is best for their context.

In fact, the draft implementation leaves as many options open as possible. If you really want to dictate links to be open in the default web browser, `open_externally=True` still does it. But the fact that `open_externally=False` is not similarly authoritarian in forbidding nothing to be opened externally is uncomfortably asymmetric - but I will accept the asymmetry. (The symmetry can be restored by introducing a default value for `open_externally` that is neither `True` or `False`, to indicate "do the default thing").
